### ISSUE TITLE
Would like to be able to manage apps in order to support add-on apps

### ISSUE THREAD

**Issue by b1conrad** [Picolab/fully-sharded-database#53]
Fundamentally, this repo implements a fully-sharded database. 

Since each entity is represented by a pico, each entity can do more than just hold the state of the entity. In fact, the `byu.hr.record` ruleset provides additional functionality, and could be thought of as an add-on app. Of course, being known to the other rulesets of the same family, it would have to be considered a _mandatory_ and _built-in_ app.

Next would come `byu.hr.manage_apps` as a built-in app to manage itself and other apps. It would come seeded with knowledge of itself (of course) and the audio app. What it would contribute is the ability to add other apps, as implemented in rulesets.

One such app might provide relationships between entities, implemented by subscriptions. These relationships could be augmented by installing the ACA-Pico rulesets to provide secure chat.

For a proof of concept, we would need two new rulesets

- `byu.hr.manage_apps`
- `byu.hr.relate` (using `io.picolabs.subscription` and the ACA-Pico rulesets)

and a slight modification to `byu.hr.core` to add a link to "Manage apps" for every pico which has that ruleset installed.

**Comment by b1conrad** [comment]
This repo uses picos as a "full-stack" for creating web applications. There is no need for a separate database, because by fully sharding, each entity is represented by a pico. 

A ruleset can share a function for each web page of the application (for the webserver tier), have rules to implement the business logic (for that tier), and use entity variables and/or `io.picolabs.pds` to store the state of the entity (for the database tier).

This issue calls for an extension of the base web application. In other words, the data can be used by multiple applications in a way similar to what can be done in a three tiered architecture.

**Comment by b1conrad** [comment]
Assuming a ruleset shares a function, say `index`, that produces a string of HTML code, it can be requested by a URL like
```
https://DOMAIN:PORT/c/ECI/query/RID/index.html
```
where portions in all caps are to be replaced by the pico engine's domain and port, the pico's event channel identifier, and the ruleset identifier, respectively.

This URL syntax, as in other RESTful services, is a path to a resource (think entity) with `c/` (think "channel") introducing the event channel identifier, `ECI/` which itself combines the _identity_ of the entity and an _authorization key_. The portions `query/`, `RID/` and `index` are expected by the pico engine so that it can provide a web page, as computed by the function `index` shared by the ruleset.

The portion ".html" signals to the pico engine _qua_ webserver that the query result is expected to have `Content-type: text/html`. For more on this see the page [Content negotiation with HTTP](https://picolabs.atlassian.net/wiki/spaces/docs/pages/358154241/Content+negotiation+with+HTTP).

**Comment by b1conrad** [comment]
Migration, or adding a new feature to a running system, without disrupting other users. The plan is to 

1. Write the `byu.hr.manage_apps` ruleset, including a function `manage` to produce its HTML page
2. Modify the `byu.hr.core` ruleset to check for the existence of that ruleset within and then generate a link to it

During testing, the manage ruleset can be installed in my own pico only, so that others won't be disturbed. As I wish others to join in testing/validation, I can install it in their picos as well. When it is ready to roll out, I can run a version of `byu.hr.patch` to place it in the pico of all entities.

**Comment by b1conrad** [comment]
Next step would be to write `byu.hr.connect` and use the "Manage apps" page to add it to my pico, and test it. Since it fundamentally involves more than just me, I would have to invite others to use their "Manage apps" page to add it as well. This would be the next step in bootstrapping.

**Comment by b1conrad** [comment]
While logged in and looking at someone else's information, if we had our own `wellKnown_Rx` available, we could then request a subscription using it, from within that other person's page. This would be done by sending the request event to the other person's pico and passing in our `wellKnown_Rx` as the attribute of that name.

So, upon logging in, let's get that ECI and store it in a cookie.

**Comment by b1conrad** [comment]
Done, in [this commit](https://github.com/Picolab/fully-sharded-database/commit/ca827d18c9fc2b253e385ad637eaca5dc1f11592), and can use it to establish a relationship, using the Subscriptions tab in another person's pico. Need to figure out the UX.

**Comment by b1conrad** [comment]
Ideally, I would be looking at another person's core information, and, since I have the `byu.hr.relate` application installed (via the manage apps UX), there would be a widget allowing me to propose a relationship with them. The widget would not appear for people who didn't have the relationship app installed in their own pico.

The widget would prompt me for:

- a description of my side of the proposed relationship (e.x. "virtual team member") sent as the "Rx_role"
- a description of their side of the proposed relationship (e.x. "team leader") sent as the "Tx_role"

The widget would have available via code:

- the ECI of their channel tagged ["wellknown_rx","tx_rx"] (obtained by code in `byu.hr.core`) sent as the "wellKnown_Tx" attribute
- values for attributes "name" and "channelType" (probably the two person names and "relationship" respectively)

The widget would send the `wrangler:subscription` event to my wellKnown_Rx from a session cookie with the above attributes. Then it would be outbound pending on my side and inbound on their side. 

Questions:

- How to surface the inbound request on their UX? So they can approve or decline.
- How would their core page know that the logged in person has the relate app? I assume a session cookie.

**Comment by b1conrad** [comment]
After today's series of commits, ending with [this commit](https://github.com/Picolab/fully-sharded-database/commit/da49dcda7e4a00872ee2c02659c309b34bdcc905), the page content looks like this:
![Screen Shot 2022-02-28 at 8 52 57 AM](https://user-images.githubusercontent.com/19273926/156014607-1d5a3a62-47e3-495d-933c-4656002a1033.png)
Next step will be to actually add an app and be able to delete it.

**Comment by b1conrad** [comment]
Instructions for creating an app:

1. Write a ruleset that shares a function which produces an HTML page for the app's UX
2. Make this function the first one shared in the `meta` block
3. Also in the `meta` block give a `name` which will be used as the button text "Manage your <name>"
4. Publish the raw ruleset source code on a server that your users can reach

Examples of names:
- for `byu.hr.manage_apps` the name is "apps"
- for `byu.hr.record` the name is "audio"

**Comment by b1conrad** [comment]
This [commit](https://github.com/Picolab/fully-sharded-database/blob/1b4eaaa4cd55085285ed240db26677dcde2c52c6/krl/byu.hr.relate.krl) of `byu.hr.relate` would be a good starting template

**Comment by windley** [comment]
Would `html.byu` be a good candidate as a general purpose library or is it too specific to what you're doing? It would be nice to have an HTML library again.

**Comment by b1conrad** [comment]
Just plain `html` would be the start of it, since `html.byu` is the start of something specific for BYU.

**Comment by b1conrad** [comment]
The plan is to have the person web page generate a `wrangler:subscription` event to the logged in person's pico. When this is being processed, the `wrangler:outbound_pending_subscription_added` event will be _raised_ in that pico (and the `wrangler:inbound_pending_subscription_added` event _sent_ to the pico of the person being observed).

The `byu.hr.relate` ruleset could simply redirect back to the page that generated the initial event. So, no AJAX would be necessary.

**Comment by b1conrad** [comment]
With [this commit](https://github.com/Picolab/fully-sharded-database/commit/660a0bc16689b15d07437385ee2446e7a4cdf8e9), correcting a typo, the relationship is offered, as a pending inbound subscription.

The next step is to produce a nicely readable invitation to accept the `byu.hr.manage_apps` and `byu.hr.relate` rulesets so that the request can be accepted. If that invitation is denied, we should just deny the subscription. How does the proposing participant know of the denial? It will have disappeared from their outbound pending list when they "Manage your relationships" for one thing. This could perhaps be reified as a piece of state in the inviting party's `byu.hr.relate` ruleset.

Haha, forgot to submit this comment yesterday! Continuing with this morning's work.

After a long series of commits, ending with [this one](https://github.com/Picolab/fully-sharded-database/commit/c82dfafab81612e96e16982a760ef950bfe101cb), we now display a message to the invitee, of the form
```
You have a request from {inviter_display_name} to acknowledge a relationship as {invitee_role} to {inviter_role}, respectively.
```

**Comment by b1conrad** [comment]
This app has no way to notify people, so the participant being invited will have to actually visit the app, and look at their own information to even see this message. People will have to communicate out of band. I have fallen into the common trap of web app developers in which they assume that people will be spending a lot of time in their app. That assumption is almost always false! Facebook, Twitter, etc. are some exceptions.

**Comment by b1conrad** [comment]
Hm. Need to know the `tags` that an app has used to define the channel to be used to request its home page. For now, I am assuming that they will be just the name of the first shared function.

**Comment by b1conrad** [comment]
After today's series of commits, ending with [this one](https://github.com/Picolab/fully-sharded-database/commit/5c9b98c80c80cf1fb0aa1a224c7a6f909384be5b), the page produced by `byu.hr.relate` looks like this:![Screen Shot 2022-03-02 at 3 47 35 PM](https://user-images.githubusercontent.com/19273926/156462734-5f754980-133d-4610-aa94-f9529cea8fbf.png)

**Comment by b1conrad** [comment]
With this series of commits (ending with this one), suggesting a relationship is better presented:
![Screen Shot 2022-03-03 at 7 01 24 AM](https://user-images.githubusercontent.com/19273926/156597265-3d0cfeba-bd5d-461e-8d3e-90a0b6e43dc1.png)
and can be deleted in the "Manage relationships" page.

**Comment by b1conrad** [comment]
Funny story. I'm basing work on links in the Manage relationships page on the v0 pico engine UI. Using the "Pico Engine UI -- How it works" [page](https://picolabs.atlassian.net/wiki/spaces/docs/pages/764084227/Pico+Engine+UI+--+How+it+works) and _found_ the template at, wait for it, line **404**.
```
404    <script id="subscriptions-template" type="text/x-handlebars-template">
``` 
404 Not Found. Haha.

**Comment by b1conrad** [comment]
Most of the recent work on this issue was about the `io.picolabs.relate` ruleset. It is intended to test the functionality of managing add-on apps.

To complete this issue as such, all that remains is allowing the participant to delete an app that they have added using the "Manage apps" page. Further work on relationships (e.x. augmenting them to be Aries connections) could then be done in a new issue.

**Comment by b1conrad** [comment]
Completed the deleting of apps with a series of commits ending [with this one](https://github.com/Picolab/fully-sharded-database/commit/d23f6668564e4cbe9de77257fca8db76c1087e73).
